src/Type/TypehintHelper.php

```diff
@@ -201,8 +201,11 @@ public static function decideType(
 			}
 
 			if (
-				(!$phpDocType instanceof NeverType || ($type instanceof MixedType && !$type->isExplicitMixed()))
-				&& $type->isSuperTypeOf(TemplateTypeHelper::resolveToBounds($phpDocType))->yes()
+				($type->isCallable()->yes() && $phpDocType->isCallable()->yes())
```

Comment:
This was the problem encountered in the original PR (at least when rebased on the current 1.11.x branch). `$type` is `callable(mixed, mixed): int` and `TemplateTypeHelper::resolveToBounds($phpDocType)` is `callable(array-key, array-key): int`. `$type->isSuperTypeOf(TemplateTypeHelper::resolveToBounds($phpDocType))` returns maybe (IMO it should return no, but I didn't look into that).

As I understand it the goal of `decideType` is to allow phpdocs to narrow down the native type (or in case of the native PHP code reflection also the signature/stub/jetbrains stuff...). The narrow-down rule probably works fine in most cases, but for callables, chances are that the `$type` may be incorrectly narrowed down too much.

I'm not sure about the consequences of this change. I'd feel better if it could only be applied to native PHP reflections, not parsed user code.